server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java

```diff
@@ -153,7 +190,7 @@ public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Set<String> i
 
         // if we are still fetching, return null to indicate it
         if (hasAnyNodeFetching(cache)) {
-            return new FetchResult<>(shardId, null, emptySet());
+            return new FetchResult<>(null, emptyMap());
```

Comment:
Have we considered building an extended version of AsyncShardFetch where we override the behavior of fetchData shard metadata? This can help easily help in doing a clean switch from batch to single shard and vice-versa. Plus it can lay foundation for other optimizations we plan to do in allocation flow. May help in avoiding code duplications as well.
Pls let me know if there are challenges in doing so. Would really like to see a `AsyncBatchShardsFetch`.

Reply:
Ya I thought of doing the same, but has its own challenges stated below:
1. `AsyncBatchShardsFetch` will take list/map of shards, and it will need to initialize the super constructor .i.e ```AsyncShardFetch``` with a single shard. How will you select that single shard, And you cant pass null because there is nonNull check. Even if pass null hypothetically, then it will pose its own challenges later for handling those checks in existing ```AsyncShardFetch```
2. We would be needing a change in FetchResult because Response structure is changing, resulting in change of function signature of fetchData hence, overruling the possibility of override. We need to change existing class (`AsyncShardFetch`) class fetchData function signature as well for that. This will make code more complex IMHO then.
3. FetchData code is almost same for batching with only change IgnoreNode handling, Overriding it will lead to duplication of code again.

Reply:
My main concern is that I see a lot of `Same as AsyncShardFetch` comments by you. Do we have a plan to completely remove single shard fetch? If we do then it should be fine to have code duplicated for now but if there's a reason that it will continue to exist then it is worth exploring options to build a batch version. 
cc: @Bukhtawar @shwetathareja